### PR TITLE
add supporting for i386 architecture.

### PR DESCRIPTION
**Pull request by ParkHanbum**
m@ubuntu:~/uftrace/tests$ ./runtest.py
Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
001 basic               : OK OK OK OK OK OK OK OK OK OK
002 argument            : OK OK OK OK OK OK OK OK OK OK
003 thread              : OK OK OK OK OK OK OK OK OK OK
004 filter_F            : OK OK OK OK OK OK OK OK OK OK
005 filter_N            : OK OK OK OK OK OK OK OK OK OK
006 filter_FN           : OK OK OK OK OK OK OK OK OK OK
007 library             : OK OK OK OK OK OK OK OK OK OK
008 daemon              : NG NG NG NG NG NG NG NG NG NG
009 fork                : OK OK OK OK OK OK OK OK OK OK
010 forkexec            : NG NG NG NG NG NG NG NG NG NG
011 vforkexec           : NG NG NG NG NG NG NG NG NG NG
012 demangle            : OK OK OK OK OK OK OK OK OK OK
013 signal              : OK OK OK OK OK OK OK OK OK OK
014 ucontext            : OK OK OK OK OK OK OK OK OK OK
015 longjmp             : OK OK OK OK OK OK OK OK OK OK
016 alloca              : OK OK OK OK OK OK NG NG NG NG
017 no_libcall          : OK OK OK OK OK OK OK OK OK OK
018 filter_regex        : OK OK OK OK OK OK OK OK OK OK
019 full_depth          : OK OK OK OK OK OK OK OK OK OK
020 filter_depth        : OK OK OK OK OK OK OK OK OK OK
021 filter_plt          : OK OK OK OK OK OK OK OK OK OK
022 filter_kernel       : SK SK SK SK SK SK SK SK SK SK
023 replay_filter       : OK OK OK OK OK OK OK OK OK OK
024 report_basic        : OK OK OK OK OK OK OK OK OK OK
025 report_s_call       : OK OK OK OK OK OK OK OK OK OK
026 filter_trigger      : OK OK OK OK OK OK OK OK OK OK
027 replay_filter_d     : OK OK OK OK OK OK OK OK OK OK
028 replay_backtrace    : OK OK OK OK OK OK OK OK OK OK
029 trigger_only        : OK OK OK OK OK OK OK OK OK OK
030 replay_trigger      : OK OK OK OK OK OK OK OK OK OK
031 filter_demangle1    : OK OK OK OK OK OK OK OK OK OK
032 filter_demangle2    : OK OK OK OK OK OK OK OK OK OK
033 filter_demangle3    : OK OK OK OK OK OK OK OK OK OK
034 filter_demangle4    : OK OK OK OK OK OK OK OK OK OK
035 filter_demangle5    : OK OK OK OK OK OK OK OK OK OK
036 replay_filter_N     : OK OK OK OK OK OK OK OK OK OK
037 trace_onoff         : OK OK OK OK OK OK OK OK OK OK
038 trace_disable       : OK OK OK OK OK OK OK OK OK OK
039 trace_onoff_F       : OK OK OK OK OK OK OK OK OK OK
040 replay_onoff        : OK OK OK OK OK OK OK OK OK OK
041 replay_onoff_N      : OK OK OK OK OK OK OK OK OK OK
042 live_disable        : OK OK OK OK OK OK OK OK OK OK
043 full_demangle       : OK OK OK OK OK OK OK OK OK OK
044 report_avg_total    : OK OK OK OK OK OK OK OK OK OK
045 report_avg_self     : OK OK OK OK OK OK OK OK OK OK
046 report_threads      : OK OK OK OK OK OK OK OK OK OK
047 signal2             : OK OK OK OK OK OK OK OK OK OK
048 malloc_impl         : OK OK OK OK OK OK OK OK OK OK
049 column_view         : OK OK OK OK OK OK OK OK OK OK
050 no_pltbind          : OK OK OK OK OK OK OK OK OK OK
051 return              : OK OK OK OK OK OK OK OK OK OK
052 nested_func         : OK OK OK OK OK NG NG NG NG NG
053 filter_time         : OK OK OK OK OK OK OK OK OK OK
054 filter_time_F       : OK OK OK OK OK OK OK OK OK OK
055 filter_time_N       : OK OK OK OK OK OK OK OK OK OK
056 filter_time_T       : OK OK OK OK OK OK OK OK OK OK
057 filter_time_D       : OK OK OK OK OK OK OK OK OK OK
058 arg_int             : OK OK OK OK OK SK SK SK SK SK
059 arg_str             : OK OK OK OK OK SK SK SK SK SK
060 arg_fmt             : OK OK OK OK OK SK SK SK SK SK
061 arg_plt             : OK OK OK OK OK OK OK OK OK OK
062 arg_char            : OK OK OK OK OK SK SK SK SK SK
063 retval              : OK OK OK OK OK SK SK SK SK SK
064 trigger_trace       : OK OK OK OK OK OK OK OK OK OK
065 arg_order           : OK OK OK OK OK SK SK SK SK SK
066 no_demangle         : OK OK OK OK OK OK OK OK OK OK
067 report_diff         : OK OK OK OK OK OK OK OK OK OK
068 filter_time_A       : OK OK OK OK OK SK SK SK SK SK
069 graph               : OK OK OK OK OK OK OK OK OK OK
070 graph_backtrace     : OK OK OK OK OK OK OK OK OK OK
071 graph_depth         : OK OK OK OK OK OK OK OK OK OK
072 no_comment          : OK OK OK OK OK OK OK OK OK OK
073 lib_filter          : OK OK OK OK OK OK OK OK OK OK
074 lib_trigger         : OK OK OK OK OK OK OK OK OK OK
075 lib_arg             : OK OK OK OK OK SK SK SK SK SK
076 lib_replay_F        : OK OK OK OK OK OK OK OK OK OK
077 lib_replay_T        : OK OK OK OK OK OK OK OK OK OK
078 max_stack           : OK OK OK OK OK OK OK OK OK OK
079 replay_kernel_D     : SK SK SK SK SK SK SK SK SK SK
080 replay_kernel_D2    : SK SK SK SK SK SK SK SK SK SK
081 kernel_depth        : SK SK SK SK SK SK SK SK SK SK
082 arg_many            : OK OK OK OK OK SK SK SK SK SK
083 arg_float           : OK OK OK OK OK SK SK SK SK SK
084 arg_mixed           : OK OK OK OK OK SK SK SK SK SK
085 arg_reg             : NG NG NG NG NG SK SK SK SK SK
086 arg_stack           : OK OK OK OK OK SK SK SK SK SK
087 arg_variadic        : OK OK OK OK OK SK SK SK SK SK
088 graph_session       : OK OK OK OK OK OK OK OK OK OK
089 graph_exit          : OK OK OK OK OK OK OK OK OK OK
090 report_recursive    : OK OK OK OK OK OK OK OK OK OK
091 replay_tid          : OK OK OK OK OK OK OK OK OK OK
092 report_tid          : OK OK OK OK OK OK OK OK OK OK
093 report_filter       : OK OK OK OK OK OK OK OK OK OK
094 report_depth        : OK OK OK OK OK OK OK OK OK OK
095 graph_tid           : OK OK OK OK OK OK OK OK OK OK
096 graph_filter        : OK OK OK OK OK OK OK OK OK OK
097 dump_basic          : OK OK OK OK OK OK OK OK OK OK
098 dump_tid            : OK OK OK OK OK OK OK OK OK OK
099 dump_filter         : OK OK OK OK OK OK OK OK OK OK
100 dump_depth          : OK OK OK OK OK OK OK OK OK OK
101 dump_chrome         : OK OK OK OK OK OK OK OK OK OK
102 dump_flamegraph     : OK OK OK OK OK OK OK OK OK OK
103 dump_kernel         : SK SK SK SK SK SK SK SK SK SK
104 graph_kernel        : SK SK SK SK SK SK SK SK SK SK
105 replay_time         : OK OK OK OK OK OK OK OK OK OK
106 report_time         : OK OK OK OK OK OK OK OK OK OK
107 dump_time           : OK OK OK OK OK OK OK OK OK OK
108 graph_time          : OK OK OK OK OK OK OK OK OK OK
109 replay_time_A       : OK OK OK OK OK SK SK SK SK SK
110 replay_time_T       : OK OK OK OK OK OK OK OK OK OK
111 kernel_tid          : SK SK SK SK SK SK SK SK SK SK
112 replay_skip         : OK OK OK OK OK OK OK OK OK OK
113 trigger_time        : OK OK OK OK OK OK OK OK OK OK
114 replay_trg_time     : OK OK OK OK OK OK OK OK OK OK
115 replay_field        : OK OK OK OK OK OK OK OK OK OK
116 field_none          : OK OK OK OK OK OK OK OK OK OK
117 time_range          : OK OK OK OK OK OK OK OK OK OK
118 thread_tsd          : OK OK OK OK OK OK OK OK OK OK
119 malloc_hook         : OK OK OK OK OK OK OK OK OK OK
120 malloc_tsd          : OK OK OK OK OK OK OK OK OK OK
121 malloc_fork         : OK OK OK OK OK OK OK OK OK OK
122 time_range2         : NG NG OK OK NG OK OK OK OK OK
123 backtrace           : OK OK OK OK OK OK OK OK OK OK
124 exception           : NG SG SG SG SG OK OK OK OK OK
125 report_range        : OK OK OK OK OK OK OK OK OK OK
126 arg_regex           : OK OK OK OK OK SK SK SK SK SK
127 arg_module          : OK OK OK OK OK SK SK SK SK SK
128 arg_module2         : OK OK OK OK OK SK SK SK SK SK
129 session_tid         : OK OK OK OK OK OK OK OK OK OK
130 thread_exec         : OK OK OK OK OK OK OK OK OK OK
131 lib_dlopen          : OK OK OK OK OK OK OK OK OK OK
132 trigger_kernel      : SK SK SK SK SK SK SK SK SK SK
133 long_string         : OK OK OK OK OK SK SK SK SK SK
134 pic_pie             : OK OK OK OK OK OK OK OK OK OK
135 trigger_time2       : OK OK OK OK OK OK OK OK OK OK
136 dynamic             : NG NG NG NG NG NG NG NG NG NG
137 kernel_tid_update   : SK SK SK SK SK SK SK SK SK SK
138 kernel_dynamic      : SK SK SK SK SK SK SK SK SK SK
139 kernel_dynamic2     : SK SK SK SK SK SK SK SK SK SK
140 dynamic_xray        : BI BI BI BI BI BI BI BI BI BI
141 recv_basic          : OK OK OK OK OK OK OK OK OK OK
142 recv_multi          : OK OK OK OK OK OK OK OK OK OK
143 recv_kernel         : SK SK SK SK SK SK SK SK SK SK
144 longjmp2            : OK OK OK OK OK OK OK OK OK OK
145 longjmp3            : OK OK OK OK OK OK OK OK OK OK
146 arg_std_string      : OK OK OK OK OK SK SK SK SK SK
147 event_sdt           : BI BI BI BI BI BI BI BI BI BI
148 event_kernel        : SK SK SK SK SK SK SK SK SK SK
149 event_kernel2       : SK SK SK SK SK SK SK SK SK SK
150 recv_event          : BI BI BI BI BI BI BI BI BI BI
151 recv_runcmd         : NG NG NG NG NG NG NG NG NG NG
152 read_proc_statm     : OK OK OK OK OK OK OK OK OK OK
153 read_page_fault     : OK OK OK OK OK OK OK OK OK OK
154 keep_pid            : NG NG NG NG NG NG NG NG NG NG
155 trigger_finish      : OK OK OK OK OK OK OK OK OK OK
156 trigger_finish2     : OK OK OK OK OK OK OK OK OK OK
157 script_python       : OK OK OK OK OK OK OK OK OK OK
158 report_diff_policy1 : OK OK OK NG OK OK OK OK OK OK
159 report_diff_policy2 : OK OK OK OK OK OK OK OK OK OK
160 report_diff_policy3 : OK OK OK OK OK OK OK OK OK OK
161 pltbind_now         : NG NG NG NG NG NG NG NG NG NG
162 pltbind_now_pie     : NG NG NG NG NG NG NG NG NG NG
163 event_sched         : NG NG NG NG NG NG NG NG NG NG
164 report_sched        : NG NG NG NG NG NG NG NG NG NG
165 graph_sched         : NG NG NG NG NG NG NG NG NG NG
166 dump_sched          : NG NG NG NG NG NG NG NG NG NG
167 recv_sched          : NG NG NG NG NG NG NG NG NG NG
168 lib_nested          : SG SG SG SG SG SG SG SG SG SG
169 script_args         : OK OK OK OK OK OK OK OK OK OK
170 script_filter       : OK OK OK OK OK OK OK OK OK OK
171 script_option       : OK OK OK OK OK OK OK OK OK OK
172 trigger_filter      : OK OK OK OK OK OK OK OK OK OK
173 trigger_args        : OK OK OK OK OK SK SK SK SK SK
174 replay_filter_kernel: SK SK SK SK SK SK SK SK SK SK
175 filter_time_read    : OK OK OK OK OK OK OK OK OK OK
176 arg_fptr            : OK OK OK OK OK OK OK OK OK OK
177 report_diff_policy4 : OK OK OK OK OK OK OK OK OK OK
178 arg_auto1           : OK OK OK OK OK OK OK OK OK OK
179 arg_auto2           : OK OK OK OK OK OK OK OK OK OK
180 arg_auto3           : OK OK OK OK OK OK OK OK OK OK

runtime test stats
====================
total  1800  Tests executed (success: 75.11%)
  OK:  1262  Test succeeded
  OK:    90  Test succeeded (with some fixup)
  NG:   149  Different test result
  NZ:     0  Non-zero return value
  SG:    14  Abnormal exit by signal
  TM:     0  Test ran too long
  BI:    30  Build failed
  LA:     0  Unsupported Language
  SK:   255  Skipped

### CODE REVIEW

**Comment by namhyung** [comment]
Wow, this is a huge contribution and I really appreciate your hard work!

But it's almost impossible to process such a gigantic patch at once.  Let's break it into pieces and I'll take a look, many thanks!

File: libmcount/plthook.c [hunk 1]
```diff
@@ -702,7 +702,7 @@ static void update_pltgot(struct mcount_thread_data *mtdp,
 }
 
 unsigned long plthook_entry(unsigned long *ret_addr, unsigned long child_idx,
-			    unsigned long module_id, struct mcount_regs *regs)
+				unsigned long module_id, struct mcount_regs *regs)
```
Comment (namhyung) [inline libmcount/plthook.c]: please don't make unnecessary whitespace changes.

File: libmcount/plthook.c [hunk 2]
```diff
@@ -744,6 +744,16 @@ unsigned long plthook_entry(unsigned long *ret_addr, unsigned long child_idx,
...
 	recursion = false;
 
+#if defined(__i386__)
+	if (child_idx > 0) {
+		if (child_idx % 8) {
+			pr_err_ns("the malformed child idx : %lx\n", child_idx);
+		}
+		child_idx = child_idx / 8;
+	}
+#endif
```
Comment (namhyung) [inline libmcount/plthook.c]: this needs some explanation - why it's invalid?

Also I'd like to avoid such ifdef-ery in the .c file as long as possible.  You may want to define a mcount_arch_XXX function for it and make them NOP for other archs.

File: misc/debug-mcount.cmd [hunk 3]
```diff
@@ -10,4 +10,3 @@ commands
 end
 
 r record -L. -d xxx --keep-pid --force tests/t-abc
-
```
Comment (namhyung) [inline misc/debug-mcount.cmd]: please don't. :)

File: tests/s-exp-mixed.c [hunk 4]
```diff
@@ -1,3 +1,40 @@
+#ifdef FASTCALL
+__attribute__((fastcall))
```
Comment (namhyung) [inline tests/s-exp-mixed.c]: I think you'd better making a new test case for the fastcall and leave this as is.

File: tests/t063_retval.py [hunk 5]
```diff
@@ -23,6 +23,7 @@ def build(self, name, cflags='', ldflags=''):
 
         return TestBase.build(self, name, cflags, ldflags)
 
+
```
Comment (namhyung) [inline tests/t063_retval.py]: not needed.

File: utils/filter.c [hunk 6]
```diff
@@ -113,8 +113,15 @@ struct uftrace_filter *uftrace_match_filter(uint64_t ip, struct rb_root *root,
 		iter = rb_entry(parent, struct uftrace_filter, node);
 
 		if (match_ip(iter, ip)) {
+#if defined(__i386__)
+			__asm__ __volatile__ ("sub $8, %esp\n");
+			__asm__ __volatile__ ("movsd %xmm0, (%esp)\n");
+#endif
 			memcpy(tr, &iter->trigger, sizeof(*tr));
```
Comment (namhyung) [inline utils/filter.c]: this is a problem as memcpy() can corrupt SSE registers as you saw.  actually libmcount uses its own memcpy implementation but this code is used both for libmcount and uftrace.

So I suggest copying by assignment instead of memcpy since we build libmcount with -mno-sse2.

```
*tr = iter->trigger;
```

File: arch/i386/mcount-arch.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+};
+
+#define HAVE_MCOUNT_ARCH_CONTEXT
+struct mcount_arch_context {
+	double xmm[ARCH_MAX_FLOAT_REGS];
+};
+
+#define FIX_PARENT_LOC
+unsigned long * mcount_arch_parent_location(struct symtabs *symtabs,
+	              							unsigned long *parent_loc,
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/i386/mcount-arch.h]: my mistake. sorry

File: arch/i386/mcount-dynamic.c [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+extern void __fentry__(void);
+extern void __xray_entry(void);
+extern void __xray_exit(void);
+
+struct xray_instr_map {
+	unsigned long addr;
+	unsigned long entry;
+	unsigned long type;
+	unsigned long count;
+};
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/i386/mcount-dynamic.c]: not yet.

File: arch/i386/mcount-support.c [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+
+    switch (spec->type) {
+        case ARG_TYPE_REG:
+            reg_idx = spec->reg_idx;
+            break;
+        default:
+            return -1; 
+    }
+
+	if (spec->type == ARG_TYPE_REG) {
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/i386/mcount-support.c]: you're right.

File: tests/s-exp-mixed.c (hunk 4 above)
Comment (ParkHanbum) [reply tests/s-exp-mixed.c]: it is need for test 085 about argument register. not fastcall.
without fastcall, x86 doesn't use register for passing argument.

File: utils/filter.c (hunk 6 above)
Comment (ParkHanbum) [reply utils/filter.c]: got it!

File: arch/i386/mcount-support.c [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+
+        pr_dbg("FIND SYMBOL\n");
+        ret_addr = *parent_loc;
+        parent_sym = find_symtabs(symtabs, ret_addr);
+        pname = symbol_getname(parent_sym, ret_addr);
+        pr_dbg("SYMBOL : %s\n", pname);
+        pr_dbg("FIND CHILD SYMBOL\n");
+        child_sym = find_symtabs(symtabs, child_ip);
+        cname = symbol_getname(child_sym, child_ip);
+        pr_dbg("SYMBOL : %s\n", cname);
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/i386/mcount-support.c]: I'll change that. thank you

File: arch/i386/mcount-support.c [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+static bool main_ret_found = false;
+
+int mcount_get_register_arg(struct mcount_arg_context *ctx, 
+                            struct uftrace_arg_spec *spec)
+{
+    struct mcount_regs *regs = ctx->regs;
+    int reg_idx;
+
+    switch (spec->type) {
+        case ARG_TYPE_REG:
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/i386/mcount-support.c]: I see. sorry.

File: libmcount/plthook.c (hunk 2 above)
Comment (ParkHanbum) [reply libmcount/plthook.c]: I do not know why, but the value is set to a multiple of 8.

**Comment by ParkHanbum** [comment]
I have corrected PR according to your advice. 
thanks you for pleasure advice!

File: tests/s-exp-int.c [hunk 12]
```diff
@@ -13,7 +13,7 @@ int int_mul(long long a, int b)
 	return a * b;
 }
 
-int int_div(unsigned a, long b)
```
Comment (ParkHanbum) [inline tests/s-exp-int.c]: this update will affect same 32bit architecture like arm.

File: utils/filter.c [hunk 13]
```diff
@@ -113,8 +113,7 @@ struct uftrace_filter *uftrace_match_filter(uint64_t ip, struct rb_root *root,
 		iter = rb_entry(parent, struct uftrace_filter, node);
 
 		if (match_ip(iter, ip)) {
-			memcpy(tr, &iter->trigger, sizeof(*tr));
-
+			*tr = iter->trigger;
```
Comment (namhyung) [inline utils/filter.c]: Please add a commit message to describe **WHY** this is needed.  Maybe you can quote (a part of) compiler error message and explanation.

File: arch/i386/mcount-dynamic.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (namhyung) [reply arch/i386/mcount-dynamic.c]: I guess they don't support i386.  If so, you should get rid of it.

[the remark this answers is not in the document]
Comment (namhyung) [reply arch/i386/mcount-dynamic.c]: Also not sure about "__fentry__" or `-mfentry` option.

File: arch/i386/mcount-support.c [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+	return 0;
+}
+
+unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
+{
+	struct sym *sym;
+
+	sym = &pd->dsymtab.sym[idx];
+	return sym->addr + ARCH_PLTHOOK_ADDR_OFFSET;
+}
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: If this is same as the default implementation, you can just drop it.

File: arch/i386/mcount-support.c [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+
+		ret_addr = *parent_loc;
+		parent_sym = find_symtabs(symtabs, ret_addr);
+		pname = symbol_getname(parent_sym, ret_addr);
+		child_sym = find_symtabs(symtabs, child_ip);
+		cname = symbol_getname(child_sym, child_ip);
+
+		// Assuming that this happens only in main.			
+		if (!strcmp(find_main[0], pname)) {
+			if (!strcmp(find_main[1], cname)) {
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: I prefer bailing out the false condition early and keep the nesting level small.  Like below?

```
if (strcmp(find_main[0], pname) || strcmp(find_main[1], cname))
    goto out;
```

File: utils/filter.c (hunk 13 above)
Comment (ParkHanbum) [reply utils/filter.c]: ok

File: arch/i386/mcount-support.c (hunk 14 above)
Comment (ParkHanbum) [reply arch/i386/mcount-support.c]: default implement is 
```
libmcount/plthook.c:671:__weak unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
libmcount/plthook.c-672-{
libmcount/plthook.c-673-        struct sym *sym;
libmcount/plthook.c-674-
libmcount/plthook.c-675-        sym = &pd->dsymtab.sym[idx];
libmcount/plthook.c-676-        return sym->addr;
libmcount/plthook.c-677-}
libmcount/plthook.c-678-
```
ARCH_PLTHOOK_ADDR_OFFSET must be addition. 
I will upload another issue about this.

File: arch/i386/mcount-support.c (hunk 15 above)
Comment (ParkHanbum) [reply arch/i386/mcount-support.c]: it's good, I'll change name of variable about pname and cname for Clarify the meaning. 
pname to parent_name, cname to called_name.
how about it?

File: arch/i386/mcount-dynamic.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply arch/i386/mcount-dynamic.c]: I'll try and drop it if can't support this.

**Comment by ParkHanbum** [comment]
@namhyung I have remove 

```
arch/i386/mcount-dynamic.c
arch/i386/mcount-event.c
arch/i386/xray.S
``` 
and I have update another point as your wish.
would you review this please?

File: arch/i386/mcount-support.c [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+		unsigned long search_ret_addr;
+
+		ret_addr = *parent_loc;
+		parent_sym = find_symtabs(symtabs, ret_addr);
+		parent_name = symbol_getname(parent_sym, ret_addr);
+		child_sym = find_symtabs(symtabs, child_ip);
+		child_name = symbol_getname(child_sym, child_ip);
+
+		// Assuming that this happens only in main..			
+		if (!(strcmp(find_main[0], parent_name) || strcmp(find_main[1], child_name))) {
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: Please break long lines.  As I said before, I prefer:

```c
if (strcmp(find_main[0], parent_name) ||
    strcmp(find_main[1], child_name))
        return parent_loc;

ret_addr = *parent_loc;
...
```

File: arch/i386/mcount-support.c [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+					parent_loc = parent_loc+i;
+					main_ret_found = true;
+				}
+			}
+		}
+		
+		// if we couldn't found correct position of return address,
+		// maybe this approach is not available anymore.
+		if (!main_ret_found) {
+			pr_dbg2("cannot find ret address of main\n");
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: then I don't want to print it repeatedly.  What about force setting to true here?

File: arch/i386/mcount.S [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+/* stack frame (with -pg): parent addr = 4(%ebp) */
+/* child addr = (%esp) */
+
+#include "utils/asm.h"
+
+GLOBAL(mcount)
+	sub $32, %esp
+	/* save registers */
+	movl %edx, 24(%esp)
+	movl %ecx, 20(%esp)
```
Comment (namhyung) [inline arch/i386/mcount.S]: I suggest save `%eax` as well just for sure.  I saw some environment passed a hidden argument via `%rax` on x86_64.

File: arch/i386/symbol.c [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+#include "uftrace.h"
+#include "utils/utils.h"
+#include "utils/symbol.h"
+
+#define R_OFFSET_POS  2
+#define JMP_INSN_SIZE 6
+#define PLTGOT_SIZE   8
+
+int arch_load_dynsymtab_bindnow(Elf *elf, struct symtab *dsymtab,
+				unsigned long offset, unsigned long flags)
```
Comment (namhyung) [inline arch/i386/symbol.c]: I think this function (or entire file) can be skipped for the first patch.

File: libmcount/plthook.c [hunk 20]
```diff
@@ -744,6 +749,8 @@ unsigned long plthook_entry(unsigned long *ret_addr, unsigned long child_idx,
 
 	recursion = false;
 
+	// for i386 architecture.
+	child_idx = mcount_arch_child_idx(child_idx);
```
Comment (namhyung) [inline libmcount/plthook.c]: Please separate this change as a new commit.

File: arch/i386/mcount-support.c [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+#define R_OFFSET_POS  2
+#define PUSH_IDX_POS  1
+#define JMP_OFS_POS   7
+#define JMP_INSN_SIZE 6
+#define PLTGOT_SIZE   8
+#define PAD_SIZE      5
+
+int mcount_arch_undo_bindnow(Elf *elf, struct plthook_data *pd)
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: This function also can be a separate commit IMHO.

**Comment by ParkHanbum** [comment]
@namhyung OK! IT IS DONE!

**Comment by namhyung** [comment]
Please rewrite the commit message header with proper format.  You'd be better reading existing commits.  Also I'd like to have well-described changelog (in the commit message).  It doesn't need to have perfect grammar but it need to describe why, how or what the commit does.  For example, "test update" is insufficient to know what is going on and why it needed.  You can say something like "it was needed because some argument spec assumed 64-bit system and failed on 32-bit systems".  Writing commit message is an important step during development of open source projects.  It should give enough context to anyone interested (at anytime).  It may be annoying at first, but you need to be familiar with it - please understand that.

Also I think the last commit should be moved before the i386 support not to break it in the first place.  I'll add more inline comment, thanks!

File: arch/i386/mcount-support.c (hunk 14 above)
Comment (namhyung) [reply arch/i386/mcount-support.c]: You need to rebase your branch onto the current master.  There were some changes in this area.

File: arch/i386/mcount-support.c [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+
+		ret_addr = *parent_loc;
+		parent_sym = find_symtabs(symtabs, ret_addr);
+		parent_name = symbol_getname(parent_sym, ret_addr);
+		child_sym = find_symtabs(symtabs, child_ip);
+		child_name = symbol_getname(child_sym, child_ip);
+
+		// Assuming that this happens only in main..			
+		if (!(strcmp(find_main[0], parent_name) || 
+		      strcmp(find_main[1], child_name))) {
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: Hmm.. why not return here?  It makes my comment below not working.

File: arch/i386/mcount-support.c (hunk 17 above)
Comment (namhyung) [reply arch/i386/mcount-support.c]: As I said, my comment assumed that it already returned for non-main functions.  Otherwise if it runs any other (e.g some initialization) function before main, this logic won't work correctly.

File: arch/i386/mcount-support.c [hunk 23]
```diff
@@ -0,0 +1,239 @@
+#include <assert.h>
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: I think this is a good place to add your authorship and copyright notation for this work. :)

File: tests/s-exp-mixed.c (hunk 4 above)
Comment (namhyung) [reply tests/s-exp-mixed.c]: In the end we need to support both fastcall and normal calls seamlessly.  Just changing the test code to use fastcall only not to fail is meaningless.  Please let this test fail and add a new test case for fastcall (ignoring other archs for now).

File: tests/t085_arg_reg.py [hunk 24]
```diff
@@ -19,7 +20,10 @@ def build(self, name, cflags='', ldflags=''):
         # cygprof doesn't support arguments now
         if cflags.find('-finstrument-functions') >= 0:
             return TestBase.TEST_SKIP
-
+        
+        if platform.machine().startswith('i686'):
+            cflags = '-march=native -m32 -mfpmath=sse -Ofast -flto -pg -O0 -D FASTCALL'
```
Comment (namhyung) [inline tests/t085_arg_reg.py]: Does all these flags are really needed?

Comment (ParkHanbum) [reply tests/t085_arg_reg.py]: yes, for forcing to use mmx registers to gcc.

Comment (ParkHanbum) [reply tests/t085_arg_reg.py]: To test passing arguments to the mmx register.

**Comment by ParkHanbum** [comment]
@namhyung I'm split testcase from this PR. 
I will split up the list of test case updates and 
send each revision and explanation to the PR.

File: arch/i386/mcount-support.c (hunk 23 above)
Comment (ParkHanbum) [reply arch/i386/mcount-support.c]: it is my honor.

**Comment by ParkHanbum** [comment]
@namhyung it is quite good for me. would you review this, please?

File: arch/i386/mcount-support.c [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+		parent_name = symbol_getname(parent_sym, ret_addr);
+		child_sym = find_symtabs(symtabs, child_ip);
+		child_name = symbol_getname(child_sym, child_ip);
+
+		// Assuming that this happens only in main..			
+		bool found_main_ret = false;
+		if (!(strcmp(find_main[0], parent_name) || 
+		      strcmp(find_main[1], child_name))) {
+			ret_addr = *parent_loc;
+			for (int i = 1; i < 5; i++ ) {
```
Comment (namhyung) [inline arch/i386/mcount-support.c]: You'd be better to use a symbol name instead of the magic number of "5" (isn't "4" enough?).  What about `MAX_STACK_ADJ` or something?

File: arch/i386/mcount-support.c (hunk 14 above)
Comment (namhyung) [reply arch/i386/mcount-support.c]: Now isn't it same? :)

File: libmcount/plthook.c (hunk 20 above)
Comment (namhyung) [reply libmcount/plthook.c]: I think this change - making `mcount_arch_child_idx()` - should be added before i386 support was enabled.  The actual implementation for i386 will be moved to the initial i386 patch.  Otherwise the initial i386 support is broken at the moment (until this commit is applied) and it'd hurt bisectability.

File: arch/i386/mcount-dynamic.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (namhyung) [reply arch/i386/mcount-dynamic.c]: I still can see the X-ray support code here.

File: arch/i386/mcount-dynamic.c [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+	unsigned xrmap_count;
+};
+
+int mcount_setup_trampoline(struct mcount_dynamic_info *mdi)
+{
+	/*
+	 * in i386, it is not possible to refer to eip.
+	 * so need a little trick. 
+	 * 
+	 * call 0x5
```
Comment (namhyung) [inline arch/i386/mcount-dynamic.c]: call 0 ?

File: arch/i386/mcount-dynamic.c [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+	unsigned long addr;
+	unsigned long entry;
+	unsigned long type;
+	unsigned long count;
+};
+
+struct arch_dynamic_info {
+	struct xray_instr_map *xrmap;
+	unsigned xrmap_count;
+};
```
Comment (namhyung) [inline arch/i386/mcount-dynamic.c]: This is not used.

File: arch/i386/mcount-dynamic.c [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+	 * jmp dword ptr [eax + 4]
+	 * 
+	 */
+	unsigned char trampoline[] = { 0xe8, 0x00, 0x00, 0x00, 0x00, 0x58, 0xff, 0x60, 0x04 };
+	unsigned long fentry_addr = (unsigned long)__fentry__;
+	struct arch_dynamic_info *adi = mdi->arch;
+	size_t trampoline_size = 16;
+
+	if (adi && adi->xrmap_count)
+		trampoline_size *= 2;
```
Comment (namhyung) [inline arch/i386/mcount-dynamic.c]: Please remove this.

File: arch/i386/mcount-dynamic.c [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+	return 0;
+}
+
+void mcount_cleanup_trampoline(struct mcount_dynamic_info *mdi)
+{
+	if (mprotect((void *)mdi->addr, mdi->size, PROT_EXEC))
+		pr_err("cannot restore trampoline due to protection");
+}
+
+void mcount_arch_find_module(struct mcount_dynamic_info *mdi)
```
Comment (namhyung) [inline arch/i386/mcount-dynamic.c]: This is to find the X-ray section and to fill the xrmap struct.  Please remove.

File: arch/i386/mcount-dynamic.c (hunk 26 above)
Comment (ParkHanbum) [reply arch/i386/mcount-dynamic.c]: @namhyung because referring eip is not possible for i386 
made a little trick which to be used for trampoline..

File: arch/i386/mcount-support.c (hunk 25 above)
Comment (ParkHanbum) [reply arch/i386/mcount-support.c]: OK!

**Comment by ParkHanbum** [comment]
@namhyung if so, I will split this "PR" to two part which "fix" and "support".
